source/microbit/microbitimage.cpp

```diff
@@ -347,27 +271,74 @@ STATIC mp_obj_t microbit_image_make_new(const mp_obj_type_t *type_in, mp_uint_t
     }
 }
 
-STATIC microbit_image_obj_t *image_crop(microbit_image_obj_t *img, mp_int_t x, mp_int_t y, mp_int_t w, mp_int_t h) {
+static void clear_rect(greyscale_t *img, mp_int_t x0, mp_int_t y0,mp_int_t x1, mp_int_t y1) {
+    for (int i = x0; i < x1; ++i) {
+        for (int j = y0; j < y1; ++j) {
+            img->setPixelValue(i, j, 0);
+        }
+    }
+}
+
+STATIC void image_blit(microbit_image_obj_t *src, greyscale_t *dest, mp_int_t x, mp_int_t y, mp_int_t w, mp_int_t h, mp_int_t xdest, mp_int_t ydest) {
     if (w < 0)
         w = 0;
     if (h < 0)
         h = 0;
-    greyscale_t *result = greyscale_new(w, h);
-    mp_int_t intersect_x0 = max(0, x);   
-    mp_int_t intersect_y0 = max(0, y);
-    mp_int_t intersect_x1 = min(img->width(), x+w);
-    mp_int_t intersect_y1 = min(img->height(), y+h);
-    /* If the cropped image is larger than the intersection then
-     * make sure that all other pixels are set to 0 */
-    if (w > intersect_x1 - intersect_x0 || h > intersect_y1 - intersect_y0) {
-        result->clear();
-    }
-    for (int i = intersect_x0; i < intersect_x1; ++i) {
-        for (int j = intersect_y0; j < intersect_y1; ++j) {
-            int val = img->getPixelValue(i, j);
-            result->setPixelValue(i-x, j-y, val);
+    mp_int_t intersect_x0 = max(max(0, x), -xdest);
+    mp_int_t intersect_y0 = max(max(0, y), -ydest);
+    mp_int_t intersect_x1 = min(min(dest->width+x-xdest, src->width()), x+w);
+    mp_int_t intersect_y1 = min(min(dest->height+y-ydest, src->height()), y+h);
+    mp_int_t xstart, xend, ystart, yend, xdel, ydel;
+    mp_int_t clear_x0 = max(0, xdest);
+    mp_int_t clear_y0 = max(0, ydest);
+    mp_int_t clear_x1 = min(dest->width, xdest+w);
+    mp_int_t clear_y1 = min(dest->height, ydest+h);
+    if (intersect_x0 >= intersect_x1 || intersect_y0 >= intersect_y1) {
+        // Nothing to copy
+        clear_rect(dest, clear_x0, clear_y0, clear_x1, clear_y1);
+        return;
+    }
+    if (x > xdest) {
+        xstart = intersect_x0; xend = intersect_x1; xdel = 1;
+    } else {
+        xstart = intersect_x1-1; xend = intersect_x0-1; xdel = -1;
+    }
+    if (y > ydest) {
+        ystart = intersect_y0; yend = intersect_y1; ydel = 1;
+    } else {
+        ystart = intersect_y1-1; yend = intersect_y0-1; ydel = -1;
+    }
+    for (int i = xstart; i != xend; i += xdel) {
+        for (int j = ystart; j != yend; j += ydel) {
+            int val = src->getPixelValue(i, j);
+            dest->setPixelValue(i+xdest-x, j+ydest-y, val);
         }
     }
+    // Adjust intersection rectange to dest
+    intersect_x0 += xdest-x;
+    intersect_y0 += ydest-y;
+    intersect_x1 += xdest-x;
+    intersect_y1 += ydest-y;
+    // Clear four rectangles in the cleared area surrounding the copied area.
+    clear_rect(dest, clear_x0, clear_y0, intersect_x0, intersect_y1);
```

Comment:
Wouldn't it be simpler to unconditionally clear the dest region first, then just copy across those pixels from the src that are in range (which might be no pixels in the case of "nothing to copy")?  This would eliminate the need for these 4 clears.